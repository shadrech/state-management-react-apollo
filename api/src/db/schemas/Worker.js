import mongoose from "mongoose";
import shortid from "shortid";

const schema = new mongoose.Schema({
  _id: {
    type: String,
    "default": shortid.generate
  },
  firstName: {
    type: String,
    trim: true
  },
  lastName: {
    type: String,
    trim: true
  },
  email: {
    type: String,
    min: 6,
    unique: true,
    required: true,
    trim: true
  },
  bio: {
    type: String,
    min: 100,
    trim: true
  },
  skills: [String],
  organization: String
});

class Worker {

  get fullName() {
    return `${this.firstName} ${this.lastName}`;
  }

}

schema.loadClass(Worker);

export default schema;

