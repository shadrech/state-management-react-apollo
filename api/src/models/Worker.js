import mongoose from "mongoose";

export function fetch(conditions = {}, options = {}) {
  return mongoose.model("Worker")
    .find(conditions, null, options);
}

export function create(fields) {
  const Worker = mongoose.model("Worker");
  const worker = new Worker({
    firstName: fields.firstName,
    lastName: fields.lastName,
    email: fields.email,
    bio: fields.bio,
    skills: fields.skills,
    organization: fields.organization
  });

  return worker.save();
}

export function update(id, fields) {
  return mongoose.model("Worker")
    .findByIdAndUpdate(id, fields, { new: true });
}

export function remove(id) {
  return mongoose.model("Worker")
    .findByIdAndRemove(id);
}
