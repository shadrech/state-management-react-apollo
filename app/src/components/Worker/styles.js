import styled from "styled-components";
import breakpoint from "styled-components-breakpoint";
import { Link } from "react-router-dom";
import Banner from "../../assets/banner.svg";

export const Worker = styled.div`
  position: relative;
  margin: 10px;
  background: white;
  width: calc(50% - 10px - 1rem);
  height: max-content;

  & > * {
    display: inline-block;
  }

  ${breakpoint('mobile', 'desktop')`
    width: calc(100% - 10px - 1rem);
  `}
`;

export const InfoWrapper = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  width: 100%;
`;

export const SkillsWrapper = styled.div`
  margin-bottom: .75rem;
`;

export const WorkerInfo = styled.div`
  padding-left: 10px;
  font-family: "resagokrregular";
  display: flex;
  align-items: center;
  position: relative;
  flex-direction: column;
  margin: 0 1.5rem;
  width: 100%;
  box-sizing: border-box;
`;

export const WorkerBio = styled(WorkerInfo)`
  width: auto;
`;

export const InfoName = styled.p`
  font-size: .9em;
  color: rgb(22, 28, 36);
  margin-bottom: 0;
  white-space: nowrap;
`;

export const InfoEmail = styled.p`
  font-size: .8em;
  margin: 0;
  color: rgb(57, 59, 61);
  white-space: nowrap;
`;

export const InfoBio = styled.p`
  font-size: .75em;
  text-align: left;
  color: rgb(57, 59, 61);
  position: relative;
  `;

export const InfoSkillTitle = styled(InfoBio)`
  font-size: .8em;
  font-weight: bolder;
  margin: 0;
  text-align: center;
`;

export const InfoSkill = styled(InfoBio)`
  display: inline-block;
  padding: 0 1.5rem;
  margin-top: 0;
  margin-bottom: 0;
  &::after {
    content: "â€¢";
    position: absolute;
    right: 0;
    font-size: 1rem;
  }
  &:last-child {
    &::after {
      content: "";
    }
  }
`;

export const WorkerButtons = styled.div`
  position: absolute;
  z-index: 50;
  right: 0;
  top: 0;
  height: 40px;
  display: flex;
  align-items: center;
`;

const Button = `
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: 50px;
  height: 50px;
  margin-left: 5px;
  &:hover {
    cursor: pointer;
    color: white;
  }
`;

export const EditButton = styled(Link)`
  ${Button}
  color: ${props => props.theme.lightGrey}
`;

export const DeleteButton = styled.div`
  ${Button}
  color: ${props => props.theme.lightGrey}
`;

export const BgImage = styled(Banner)`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 40px;
`;

export const ProfileImage = styled.img`
  border-radius: 50%;
  top: 15px;
  z-index: 10;
  position: relative;
  height: 65px;
  width: 65px;
  border: 4px solid white;
`;

export const Divider = styled.p`
  position: relative;
  display: block;
  margin: .5rem .5rem 0;
  text-align: left;
  width: 100%;
  font-size: .8rem;
  color: ${props => props.theme.darkBlue};
  i {
    margin: 0 .5rem;
    font-size: .8rem;
    ${props => props.isOpened && `
      transform: rotate(180deg);
    `}
  }
  &::after {
    content: "";
    height: 1px;
    background: grey;
    width: calc(100% - 6.5rem);
    position: absolute;
    top: 50%;
    transform: scaleY(.4);
    opacity: .4;
  }
  &:hover {
    cursor: pointer;
  }
`;
