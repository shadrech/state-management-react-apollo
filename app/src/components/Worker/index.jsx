import React from "react";
import { useMutation } from '@apollo/react-hooks';
import { Collapse } from "react-collapse";
import { confirmAlert } from "react-confirm-alert";
import "react-confirm-alert/src/react-confirm-alert.css";
import { Worker, WorkerButtons, EditButton, DeleteButton, WorkerInfo, InfoEmail, InfoName, BgImage, ProfileImage, WorkerBio, InfoBio, InfoWrapper, InfoSkill, InfoSkillTitle, Divider, SkillsWrapper } from "./styles";
import setWorkerOpenMutation from "../../graphql/mutations/setWorkerOpen.graphql";

const WorkerComponent = props => {
  const { worker } = props;
  const deleteWorker = () => confirmAlert({
    title: "Confirm delete",
    message: `Are you sure you want to delete ${worker.firstName}?`,
    confirmLabel: "Yes",
    cancelLabel: "No",
    onConfirm: () => console.log("You're fired!"),
    onCancel: () => console.log("Not today")
  });
  const rand = Math.random();
  const [setWorkerOpen] = useMutation(setWorkerOpenMutation);

  return (
    <Worker>
      <BgImage />
      <WorkerButtons>
        <EditButton to={`/worker/update/${worker.id}`}><i className="fa fa-pencil" aria-hidden="true"></i></EditButton>
        <DeleteButton onClick={deleteWorker}><i className="fa fa-trash" aria-hidden="true"></i></DeleteButton>
      </WorkerButtons>
      <InfoWrapper>
        <WorkerInfo>
          <ProfileImage alt="Random Profile" src={`https://randomuser.me/api/portraits/${rand > 0.5 ? "women" : "men"}/${Math.floor(rand * 100)}.jpg`} />
          <InfoName>{`${worker.firstName} ${worker.lastName}`}</InfoName>
          <InfoEmail>{worker.email}</InfoEmail>
        </WorkerInfo>
      </InfoWrapper>
      <Divider isOpened={worker.isOpen} onClick={() => {
        setWorkerOpen({ variables: {
          id: worker.id,
          isOpen: !worker.isOpen
        } });
      }}>
        view more
        <i className="fas fa-angle-down"></i>
      </Divider>
      <Collapse isOpened={worker.isOpen}>
        <WorkerBio>
          <InfoBio>{worker.bio}</InfoBio>
          <SkillsWrapper>
            <InfoSkillTitle>Skills</InfoSkillTitle>
            {worker.skills.map(skill => <InfoSkill key={skill}>{skill}</InfoSkill>)}
          </SkillsWrapper>
        </WorkerBio>
      </Collapse>
    </Worker>
  );
}

export default React.memo(WorkerComponent);
