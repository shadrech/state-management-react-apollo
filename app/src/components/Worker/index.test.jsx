import React from "react";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import { MockedProvider } from "@apollo/react-testing";
import { InMemoryCache } from 'apollo-boost';
import Worker from ".";
import setWorkerOpenMutation from "../../graphql/mutations/setWorkerOpen.graphql";

const defaultMocks = [{
  request: {
    query: setWorkerOpenMutation,
    variables: { id: 1, isOpen: true },
  },
  result: () => null,
}];

const worker = {
  id: 1,
  firstName: 'Tatenda',
  lastName: 'Chawanzwa',
  email: 'tatenda@chadigital.com',
  isOpen: false,
  bio: 'My short life story blah blah blah',
  skills: ['Node', 'React', 'Basketball', 'Speaker']
}

function renderComponent(component, { mocks = defaultMocks, cache, resolvers } = {}) {
  return mount(
    <MemoryRouter>
      <MockedProvider mock={mocks} cache={cache || new InMemoryCache()} resolvers={resolvers}>
        {component}
      </MockedProvider>
    </MemoryRouter>
  )
}

describe('Worker', () => {

  beforeEach(() => {
    jest.spyOn(Math, 'random').mockImplementation(() => 0.6);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render children as expected', () => {
    const component = renderComponent(<Worker worker={worker} />);
    expect(component.find('Memo(WorkerComponent)').html()).toMatchSnapshot();
  });

  test('should fire the setWorkerOpenMutation when Divider is clicked', () => {
    const mutationFn = jest.fn();
    jest.spyOn(ApolloHooks, 'useMutation').mockImplementation(() => ([mutationFn]));
    const component = mount(
      <MemoryRouter>
        <Worker worker={worker} />
      </MemoryRouter>
    );
    expect(ApolloHooks.useMutation).toHaveBeenCalledWith(setWorkerOpenMutation);
    component.find('StyledComponent[data-test-id="worker-divider"]').simulate('click');
    expect(mutationFn).toHaveBeenCalledTimes(1);
    expect(mutationFn).toHaveBeenCalledWith({ variables: {
      id: 1,
      isOpen: true
    } });
  });

});
