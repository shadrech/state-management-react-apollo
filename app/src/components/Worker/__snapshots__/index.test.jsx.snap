// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Worker should render as expected 1`] = `"<div class=\\"sc-bdVaJa gxXPlU\\"><svg class=\\"sc-VigVT idhQgL\\" data-file-name=\\"SvgBanner\\"></svg><div class=\\"sc-iwsKbI bexaRc\\"><a class=\\"sc-gZMcBi capmkh\\" href=\\"/worker/update/1\\"><i class=\\"fa fa-pencil\\" aria-hidden=\\"true\\"></i></a><div class=\\"sc-gqjmRU JVQoM\\"><i class=\\"fa fa-trash\\" aria-hidden=\\"true\\"></i></div></div><div class=\\"sc-bwzfXH fsIiXN\\"><div class=\\"sc-bxivhb iyqMsS\\"><img alt=\\"Random Profile\\" src=\\"https://randomuser.me/api/portraits/women/60.jpg\\" class=\\"sc-jTzLTM jlRCVJ\\"><p class=\\"sc-EHOje bqzkLF\\">Tatenda Chawanzwa</p><p class=\\"sc-bZQynM cvCSUk\\">tatenda@chadigital.com</p></div></div><p data-test-id=\\"worker-divider\\" class=\\"sc-fjdhpX hyhPRN\\">view more<i class=\\"fas fa-angle-down\\"></i></p><div class=\\"ReactCollapse--collapse\\" style=\\"overflow: hidden; height: 0px;\\"><div class=\\"ReactCollapse--content\\"><div class=\\"sc-bxivhb sc-ifAKCX kYthvX\\"><p class=\\"sc-gzVnrw dzCpZc\\">My short life story blah blah blah</p><div class=\\"sc-htpNat kEJEQX\\"><p class=\\"sc-gzVnrw sc-htoDjs iMSJeQ\\">Skills</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">Node</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">React</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">Basketball</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">Speaker</p></div></div></div></div></div>"`;

exports[`Worker should render children as expected 1`] = `"<div class=\\"sc-bdVaJa gxXPlU\\"><svg class=\\"sc-VigVT idhQgL\\" data-file-name=\\"SvgBanner\\"></svg><div class=\\"sc-iwsKbI bexaRc\\"><a class=\\"sc-gZMcBi capmkh\\" href=\\"/worker/update/1\\"><i class=\\"fa fa-pencil\\" aria-hidden=\\"true\\"></i></a><div class=\\"sc-gqjmRU JVQoM\\"><i class=\\"fa fa-trash\\" aria-hidden=\\"true\\"></i></div></div><div class=\\"sc-bwzfXH fsIiXN\\"><div class=\\"sc-bxivhb iyqMsS\\"><img alt=\\"Random Profile\\" src=\\"https://randomuser.me/api/portraits/women/60.jpg\\" class=\\"sc-jTzLTM jlRCVJ\\"><p class=\\"sc-EHOje bqzkLF\\">Tatenda Chawanzwa</p><p class=\\"sc-bZQynM cvCSUk\\">tatenda@chadigital.com</p></div></div><p data-test-id=\\"worker-divider\\" class=\\"sc-fjdhpX hyhPRN\\">view more<i class=\\"fas fa-angle-down\\"></i></p><div class=\\"ReactCollapse--collapse\\" style=\\"overflow: hidden; height: 0px;\\"><div class=\\"ReactCollapse--content\\"><div class=\\"sc-bxivhb sc-ifAKCX kYthvX\\"><p class=\\"sc-gzVnrw dzCpZc\\">My short life story blah blah blah</p><div class=\\"sc-htpNat kEJEQX\\"><p class=\\"sc-gzVnrw sc-htoDjs iMSJeQ\\">Skills</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">Node</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">React</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">Basketball</p><p class=\\"sc-gzVnrw sc-dnqmqq iTTTiO\\">Speaker</p></div></div></div></div></div>"`;
